server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -315,6 +315,11 @@ public class IndexShard extends AbstractIndexShardComponent implements IndicesCl
     // the translog keeps track of the GCP, but unpromotable shards have no translog so we need to track the GCP here instead
     private volatile long globalCheckPointIfUnpromotable;
 
+    /**
+     * Indicates that the {@link #close(String, boolean, Executor, ActionListener)} has been called
+     */
+    private final AtomicBoolean isClosing = new AtomicBoolean();
```

Comment:
I wonder why we needed to introduce this new flag?

Reply:
Only to help with some Engine.Warmer logic that is executed by the refresh listener and cannot abort early if the shard is closing.